store.rb

```diff
@@ -246,13 +264,12 @@ def reload_instances()
     # we set both cache and the timestamp of that version fetched
     # Since timestamp will be monotonically increasing, we are
     # sure that cache set will always have newer versions
-
     fetch_start_time = Time.now
-    instances = load_instances_from_zk.freeze
+
     @cache_mutex.synchronize do
-      if fetch_start_time > @cache_results_last_fetched_time then
-        @cache_results_last_fetched_time = fetch_start_time
-        @cache_results = instances
+      if fetch_start_time > @cache_results_last_fetched_time
```

Comment:
the only way this condition can be false is if multiple threads are stuck behind the `synchronize`, and when a thread is allowed to proceed, it is randomly chosen to be a newer thread than some older threads that are also stuck. that's pretty confusing. it's also what's responsible for your load reduction -- in the old model, multiple threads could talk to ZK, but in the new model only a single thread is allowed to do so. but in your new model, multiple threads which are stuck will still talk to ZK, even if an earlier stuck thread has already gotten all of the updates.

what behavior do you want? i think you want to `load_instances` if the reload interval has come in the background thread, or if there are unprocessed updates from the watchers. you also want only one thread to talk to ZK, to effect load reduction. you also additionally want a fetch to clear all pending fetches. to make this more clear, i propose the following:

in the event handler for the watchers (in `setup_watchers`), instead of calling `reload_instances`, you should do:
```ruby
do |event|
  @log.info "Children added/deleted"
  @updates_from_watchers_are_pending = true
  reset_watchers_on_zk
end
```

in the loop for the fetch thread (`start_fetch_thread`), do the following:
```ruby
while @cache_fetch_thread_should_run do
  begin
    now = Time.now
    if @last_fetch_time + @cache_fetch_interval < now || @updates_from_watches_are_pending
      @updates_from_watchers_are_pending = false
      @last_fetch_time = new
      @cache_results = load_instances_from_zk.freeze
   else
     sleep(FETCH_THREAD_LOOP_INTERVAL)
   end
  end
end
```

with `FETCH_THREAD_LOOP_INTERVAL` being suitably small, like `0.1`. this entire function could then be deleted. i predict this will have an additional small load reduction over the existing proposal.